### PR TITLE
Performance and authentication fixes.

### PR DESCRIPTION
**Pull request by begoldsm**
This PR addresses issues with performance for very large folder upload and download (issue: #151)
NOTE: Very large folders of small files will still be relatively slow due to latency limitations of per request throughput and file I/O for directory enumeration on the client (for upload) and the server (for download).

This pr also addresses an issue with token expiration and refresh using the built in auth() method that was discovered during investigation of the large folder performance issue.

### CODE REVIEW

File: azure/datalake/store/core.py [hunk 1]
```diff
@@ -149,21 +151,28 @@ def _walk(self, path):
...
         prefix = path.globless_prefix
-        allfiles = self.walk(prefix)
+        allfiles = self.walk(prefix, details)
         if prefix == path:
             return allfiles
-        return [f for f in allfiles if AzureDLPath(f).match(path.as_posix())]
+        if details:
+            return [f for f in allfiles if AzureDLPath(f['name']).match(path_as_posix)]
+        
+        return [f for f in allfiles if AzureDLPath(f).match(path_as_posix)]
```
Comment (lmazuel) [inline azure/datalake/store/core.py]: to avoid redondancy, I would prefer:
```python
return [f for f in allfiles if AzureDLPath(f['name'] if details else f).match(path_as_posix)]
```

same above in walk

File: azure/datalake/store/core.py [hunk 2]
```diff
@@ -249,13 +259,20 @@ def _acl_call(self, action, path, acl_spec=None):
...
+            optionally indicates that the cache of files should be invalidated after this operation
+            This should always be done for set and remove operations, since the state of the file or folder has changed.
         """
         parms = {}
         path = AzureDLPath(path).trim()
         if acl_spec:
             parms['aclSpec'] = acl_spec
         
-        return self.azure.call(action, path.as_posix(), **parms)
+        to_return = self.azure.call(action, path.as_posix(), **parms)
```
Comment (lmazuel) [inline azure/datalake/store/core.py]: just confirm, `path.as_posix` is fast enough so you can call it several time without a variable?

Comment (begoldsm) [inline azure/datalake/store/core.py]: It should be quick but, within a method, if I am calling it more than once I would want to assign it to a variable. If this is somewhere I am calling it more than once I will update it.

File: azure/datalake/store/core.py (hunk 1 above)
Comment (begoldsm) [inline azure/datalake/store/core.py]: Sounds good will do (and didn't know I could do this, fun!)

File: azure/datalake/store/lib.py [hunk 3]
```diff
@@ -157,12 +129,37 @@ def __init__(self, token):
...
+    
+    def refresh_token(self, authority=None):
+        """ Refresh an expired authorization token
+
+        Parameters
+        ----------
+        authority: string
+            The full URI of the authentication authority to authenticate against (such as https://login.microsoftonline.com/)
+        """
+        if self.token.get('refresh', False) is False:
```
Comment (brettcannon) [inline azure/datalake/store/lib.py]: Is there a need to explicitly test against False? Would None or the empty string be valid to not raise this exception?

File: tests/test_core.py [hunk 4]
```diff
@@ -206,6 +206,16 @@ def test_glob_walk(azure):
 
         assert set(azure.walk(test_dir / 'c')) == set(azure.walk(test_dir / 'c'))
 
+        # test glob and walk with details=True
+        glob_details = azure.glob(test_dir / '*', details=True)
+
+        # validate that the objects are subscriptable
+        assert glob_details[0]['name'] != None
```
Comment (brettcannon) [inline tests/test_core.py]: You want to use `is not None`.

File: azure/datalake/store/lib.py (hunk 3 above)
Comment (begoldsm) [inline azure/datalake/store/lib.py]: This was just a lift and shift of the existing method, but the reasoning is that we are trusting the calling class, in this case, only lib.auth() should ever construct one of these, but I am not sure how to prevent this class from being used by an end user. Basically, if there is no 'refresh' key in the token dictionary, then the token can't be refreshed. If there is a key, we will try to refresh the token with whatever the value is in that key. If the value is bad (or none or whatever), then refresh will fail at that point.

File: tests/test_core.py (hunk 4 above)
Comment (begoldsm) [inline tests/test_core.py]: done!

File: azure/datalake/store/lib.py (hunk 3 above)
Comment (brettcannon) [inline azure/datalake/store/lib.py]: Start its name with an `_` and then you're fine. 😁

Comment (brettcannon) [inline azure/datalake/store/lib.py]: Anyway, my point is that the `False` requirement just seems overly strict for detecting a failure case as this will let `None` be considered a refresh token which isn't accurate (nor is the empty string).
